packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_InsertGlobalIndices_FE_SP.hpp

```diff
@@ -272,6 +277,226 @@ int executeInsertGlobalIndicesFESP_(const Teuchos::RCP<const Teuchos::Comm<int>
+  //
+  //    +-----+-----+-----+-----+
+  //    |  2  | -1  |     | -1  |
+  //    +-----+-----+-----+-----+
+  //    | -1  |  2  | -1  |     |
+  //    +-----+-----+-----+-----+
+  //    |     | -1  |  2  | -1  |
+  //    +-----+-----+-----+-----+
+  //    | -1  |     | -1  |  2  |
+  //    +-----+-----+-----+-----+
+  //
+  // For Type 1 matrix fill, we create the fe_matrix object and will fill it
+  // in the same manner as we filled in the graph but in this case, nodes
+  // associated with each element will receive contributions according to
+  // the row in this stencil.
+  //
+  // In this example, the calls to sumIntoGlobalValues() on 1 core will look like:
+  //   Element 0
+  // - sumIntoGlobalValues( 0,  [  0  1  5  4  ],  [  2  -1  0  -1  ])
+  // - sumIntoGlobalValues( 1,  [  0  1  5  4  ],  [  -1  2  -1  0  ])
+  // - sumIntoGlobalValues( 5,  [  0  1  5  4  ],  [  0  -1  2  -1  ])
+  // - sumIntoGlobalValues( 4,  [  0  1  5  4  ],  [  -1  0  -1  2  ])
+  // Element 1
+  // - sumIntoGlobalValues( 1,  [  1  2  6  5  ],  [  2  -1  0  -1  ])
+  // - sumIntoGlobalValues( 2,  [  1  2  6  5  ],  [  -1  2  -1  0  ])
+  // - sumIntoGlobalValues( 6,  [  1  2  6  5  ],  [  0  -1  2  -1  ])
+  // - sumIntoGlobalValues( 5,  [  1  2  6  5  ],  [  -1  0  -1  2  ])
+  // Element 2
+  // - sumIntoGlobalValues( 2,  [  2  3  7  6  ],  [  2  -1  0  -1  ])
+  // - sumIntoGlobalValues( 3,  [  2  3  7  6  ],  [  -1  2  -1  0  ])
+  // - sumIntoGlobalValues( 7,  [  2  3  7  6  ],  [  0  -1  2  -1  ])
+  // - sumIntoGlobalValues( 6,  [  2  3  7  6  ],  [  -1  0  -1  2  ])
+  RCP<TimeMonitor> timerElementLoopMemory = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("3.1) ElementLoop  (Memory)")));
+
+  RCP<fe_matrix_t> fe_matrix = rcp(new fe_matrix_t(fe_graph));
+  RCP<fe_multivector_t> rhs =
+    rcp (new fe_multivector_t(domain_map, fe_graph->getImporter(), 1));
+
+  auto localMatrix  = fe_matrix->getLocalMatrix();
+  auto localRHS     = rhs->getLocalViewDevice();
+  auto localMap     = owned_plus_shared_map->getLocalMap();
+  auto localColMap  = fe_matrix->getColMap()->getLocalMap();
+
+  // Because we're processing elements in parallel, we need storage for all of them
+  int numOwnedElements = mesh.getNumOwnedElements();
+  int nperel = owned_element_to_node_ids.extent(1); 
+  pair_type alln = pair_type(0,nperel);
+  scalar_2d_array_t all_element_matrix("all_element_matrix",nperel*numOwnedElements);
+  scalar_1d_array_t all_element_rhs("all_element_rhs",nperel*numOwnedElements);
+  local_ordinal_view_t  all_lcids("all_lids",nperel*numOwnedElements);
+
+  timerElementLoopMemory=Teuchos::null;
+  RCP<TimeMonitor> timerElementLoopMatrix = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("3.2) ElementLoop  (Matrix)")));
+
+  // Loop over elements
+  Tpetra::beginFill(*fe_matrix,*rhs);
+  Kokkos::parallel_for(Kokkos::RangePolicy<execution_space_t>(0, numOwnedElements),KOKKOS_LAMBDA(const size_t& element_gidx) {
```

Comment:
1. Best practice is to give all `Kokkos::parallel_*` a label (as a string first argument).
2. `numOwnedElements` is an `int`, but you use `size_t` as the loop index.
3. There's no value in passing `size_t` by const reference.  Just pass it by value.

Reply:
@GeoffDanielson More feedback :)